src/vs/vscode.proposed.d.ts

```diff
@@ -1554,24 +1561,72 @@ declare module 'vscode' {
 	}
 
 	export interface NotebookProvider {
-		resolveNotebook(editor: NotebookEditor): Promise<void>;
-		executeCell(document: NotebookDocument, cell: NotebookCell | undefined): Promise<void>;
+		resolveNotebook(editor: NotebookEditor): Promise<void>; // rchiodo: Not sure what this is for? Notebook is loaded? Like prime the server? Or does this somehow generate the document?
 		save(document: NotebookDocument): Promise<boolean>;
 	}
 
+	/**
+	 * Result of an executeCell.
+	 */
+	export interface NotebookCellExecutionResult {
+		/**
+		 * Event fired when new cell output comes through
+		 */
+		readonly onCellOutput: Event<CellOutput[]>;
+		/**
+		 * Event fired when output should be cleared. Boolean flag indicates if should wait for next onOutput for clear.
+		 */
+		readonly onClear: Event< boolean>;
+		/**
+		 * Event fired when document wide output comes through
+		 */
+		readonly onDocumentOutput: Event<CellOutput[]>;
+		/**
+		 * Event fired when execution is complete
+		 */
+		readonly onComplete: Event<void>;
+	}
+
+	/**
+	 * Allows execution of notebook cells.
+	 */
+	export interface NotebookExecution {
+		readonly executionCount: number;
+		executeCell(document: NotebookDocument, cell: NotebookCell): NotebookCellExecutionResult;
+	}
+
+	/**
+	 * Allows observing the execution of cells from a NotebookExecution
+	 */
+	export interface NotebookExecutionObserver {
+		onExecuteCell(document: NotebookDocument, cell: NotebookCell, eventProvider: NotebookCellExecutionResult): void;
+	}
+
 	export interface NotebookOutputSelector {
 		type: string;
 		subTypes?: string[];
 	}
 
+	/**
+	 * Allows transformation of output from raw execution. Called during execution.
+	 */
+	export interface NotebookOutputTransformer {
+		/**
+		 *
+		 * @returns the next item in the chain
+		 *
+		 */
+		transformOutput(document: NotebookDocument, cell: NotebookCell, output: CellOutput): CellOutput;
+	}
+
 	export interface NotebookOutputRenderer {
 		/**
 		 *
 		 * @returns HTML fragment. We can probably return `CellOutput` instead of string ?
 		 *
 		 */
 		render(document: NotebookDocument, cell: NotebookCell, output: CellOutput, mimeType: string): string;
-		preloads?: Uri[];
+		preloads?: Uri[]; // rchiodo: What is this for?
```

Comment:
To preload `react` or other libraries that may be required when rendering the output.

Reply:
Ah so they woudln't have to be in the returned string

Reply:
Yes. this way the preload is a one time task and loading of resources (first time) is delegated to VS Code.